GSoC2018 report generator

Trodrige opened the pull request:
Ability to connect a second database (report generator database), and use.

Trodrige committed:
Adding migrations and models for report generator tables

Trodrige committed:
An attempt to add Factories and seeders for the report generator. Added a report generator table connection

aethelwulffe commented:
@pri2si17-1997 Hit it dude! :)

Trodrige committed:
Added the junction table for report format and draggable component relationship. Updated instructions

Trodrige committed:
display draggable components

aethelwulffe commented on Documentation/instructions.txt:
```diff
@@ -57,7 +57,8 @@ Instructions:
 
 12. You can seed the database width
-    'php artisan db:seed'
+    'php artisan db:seed' // This shouldn't work for now.
```
This is obviously a WIP comment.  Making a comment here just so we make sure to check for these when it is final review time.

aethelwulffe commented on Modules/ReportGenerator/Database/Migrations/2018_06_04_083903_create_draggable_component_report_format_table.php:
```diff
@@ -0,0 +1,44 @@
+<?php
```
I assume these names are for WIP versioning for now, or are they intended to serve in a meaningful way to version components later?

Trodrige replied:
@aethelwulffe that's the Laravel convention. I am trying to make the table names as obvious as possible.

aethelwulffe commented on Modules/ReportGenerator/Entities/DraggableComponent.php:
```diff
@@ -32,4 +33,14 @@ class DraggableComponent extends Model
+    */
+    public function report_formats()
+    {
+        return $this->belongsToMany('ReportFormat')->withTimestamps();
```
I like the naming convention for many-to-one, one-to-many elements.  Nice.

Trodrige replied:
That's the beauty of Laravel!

aethelwulffe commented on Modules/ReportGenerator/Resources/views/index.blade.php:
```diff
@@ -81,7 +81,7 @@
                         <div id="collapseOne" class="collapse show" aria-labelledby="headingOne" data-parent="#accordion">
                             <div class="card-body" id="second">
-                                <p class="note">Why am I stil empty?</p>
+                                <p class="note">Why am I still empty?</p>
```
I feel the same way.  😸

aethelwulffe replied:
Why is my still empty? 🍷

aethelwulffe commented on public/assets/css/master.css:
```diff
@@ -6,7 +6,7 @@
 }
 
-#draggable{
+.draggable{
```
ID to class?  Was that an error correction or just a refactor of the approach?

aethelwulffe reviewed (commented):
Nice work.  Much love here.

pri2si17-1997 commented on database/migrations/2018_05_18_085927_create_list_options_table.php:
```diff
@@ -0,0 +1,49 @@
+            $table->boolean('is_default')->default(0)->comment = "0 -> False, 1 -> True.";
+            $table->float('option_value', 8, 2)->default(0)->comment = "";
+            $table->string('mapping', 31)->comment = "";
+            $table->string('notes', 255)->comment = "This stores the meaning or usefulness of the component.";
```
Limiting notes to string is not a good idea. You can use `text` instead. Notes may exceed 255 char.

pri2si17-1997 commented on Modules/ReportGenerator/Database/Migrations/2018_05_22_153827_create_draggable_components_table.php:
```diff
@@ -0,0 +1,49 @@
+            $table->string('title', 255)->comment = "This is the text on the component that end users see.";
+            $table->integer('order', 0)->comment = "The order in which components appear in the list.";
+            $table->boolean('active')->default(1)->comment = "0 -> False, 1 -> True whether the component should be active or not.";
+            $table->json('notes', 255)->comment = "This stores the fields that the component relates to in the database.";
```
Why `notes` to be of json type? I can't understand from your comment. Any example will be appreciated.

pri2si17-1997 commented on Modules/ReportGenerator/Database/Migrations/2018_05_25_203256_create_report_formats_table.php:
```diff
@@ -0,0 +1,43 @@
+            $table->string('title', 255)->comment = "This is the report name e.g Patient List.";
+            $table->string('system_feature', 255)->comment = "The system feature under which the report belongs.";
+            $table->text('description')->comment = "This describes the report format briefly.";
+            $table->json('draggable_components_id')->comment = "This stores the id of all the components that belong to this report format";
```
Not able to get `draggable_components_id` field's type.

pri2si17-1997 commented:
This is a general feedback looking at one go. I would test your PR explicitly and let you know @Trodrige .

Sorry @aethelwulffe , been very busy last week, with academic major and all.

teryhill commented:
![image](https://user-images.githubusercontent.com/4655366/41070084-ee5b0026-69be-11e8-879e-405f26e54767.png)

Commit:
Getting the option IDs of selected components

aethelwulffe commented:
I assume this is open elsewhere now?
I got a data set, and I wannaa play with it.
Evals are due ya know. :)

Trodrige commented:
@aethelwulffe I'm sorry for the inconvenience. I renamed the branch to show 2018 instead of 2017. I didn't realize this happened.

Trodrige commented:
@aethelwulffe please let me know if everything is fine now

aethelwulffe commented:
Looks good to me!
I just want to play with the code.  I don't (as much) much care what it is called!

Commit:
Sending CSRF header and getting response from controller

pri2si17-1997 commented:
Hi @Trodrige , You need to add new instructions, from scratch, how to handle two databases. It would be easier to follow.

aethelwulffe commented on Modules/ReportGenerator/Http/Controllers/ReportGeneratorController.php:
```diff
@@ -33,14 +33,25 @@ public function index()
-      return url('reportgenerator/showReport');
+      $option_ids = serialize($option_ids);
+      return response()->json([
+          'redirecturl' => 'http://localhost:8000/reportgenerator/report/'.$option_ids,
```
Shoutcast port? :)

aethelwulffe replied:
Not sure what this affects, but might need to be configurable/discoverable.  For instance, this would not work in my production setup.  localhost does one thing.  127.0.0.1 does something completely different.
-I am sure you have this under consideration, or it is actually a non-issue...I am just making sure you know I am here!  😺

Trodrige replied:
I have this @aethelwulffe I'll use the $DB_HOST global.

aethelwulffe commented on Modules/ReportGenerator/Resources/views/layouts/master.blade.php:
```diff
@@ -1,11 +1,12 @@
         <meta http-equiv="X-UA-Compatible" content="IE=edge">
         <meta name="viewport" content="width=device-width, initial-scale=1">
         <meta name="author" content="Tigpezeghe Rodrige K. @ tigrodrige@gmail.com">
-        <metan name="description" content="GSoC2018: Building a report generator for LibreHealthEHR">
```
oops

Trodrige replied:
Taken care of!

aethelwulffe commented on public/assets/js/master.js:
```diff
@@ -43,19 +43,32 @@ function generate(IDs) {
     success: function(response){
-      alert("You've successfully sent the unique ids");
+      console.log(response);
+      //alert("You've successfully sent the unique ids");
```
Debug?

Trodrige replied:
Yes, debug.

aethelwulffe commented on public/assets/js/master.js:
```diff
@@ -43,19 +43,32 @@ function generate(IDs) {
+  $('#generate-button').click(function(e) {
+    e.preventDefault();
     if (jQuery.isEmptyObject(IDs)) { // check if the user has dropped any component before trying to generate a report
       alert("Please drag and drop components before you can generate a report!");
```
"Cannot create a blank report.  Please drag and drop at least one component into your new report before attempting to generate it."
OR
"Cannot generate a blank report!  You must add at least one component."

Trodrige replied:
I thought of that, and it escaped me. Handling it now!

pri2si17-1997 commented on Modules/ReportGenerator/Database/Migrations/2018_05_25_203256_create_report_formats_table.php:
```diff
@@ -0,0 +1,46 @@
+<?php
+/**
+ * This file creates draggable_components for the report-generator.
```
It should be "This file created report_format for report-generator." . Just a guess.

Trodrige replied:
Checked!

pri2si17-1997 commented on Modules/ReportGenerator/Database/Migrations/2018_06_04_083903_create_draggable_component_report_format_table.php:
```diff
@@ -0,0 +1,44 @@
+            $table->foreign('draggable_component_id')->references('id')->on('draggable_components');
+            $table->integer('report_format_id')->unsigned()->comment = "Ussed to access the report_formats.";
+            $table->foreign('report_format_id')->references('id')->on('report_formats');
+
```
There should be something, that what happens if it is deleted. Like `onDelete('cascade')`.

Trodrige replied:
checked!

pri2si17-1997 commented on Modules/ReportGenerator/Resources/views/index.blade.php:
```diff
@@ -96,31 +101,14 @@
-                                    <div class="col-sm-3" id="draggable"><p>I am seven</p></div>
-                                    <div class="col-sm-3" id="draggable"><p>I am eight</p></div>
+                                <div class="row" id="draggable-column">
+                                    @foreach ($draggableComponents as $draggableComponent)
```
This is good use of foreach to write compact code. :+1:

aethelwulffe replied:
foreach is indeed a beautiful construct.  One that sometimes leaves me thinking "My code can't work!  There isn't enough....CODE!"  😜

Trodrige replied:
My mentors catch everything!!!

Trodrige committed:
corrected previous PR. Added command to create two laravel-ehr databases

aethelwulffe commented:
Pri is your mentor here.  I am the Laravel noob peanut gallery blowing the Vuvuzela.

Trodrige committed:
An attempt to get data from selected columns, and send to views

Trodrige commented:
Please don't look at the last commit for now! I'm still working on it! Thanks!

Trodrige committed:
Attempt to retrieve data

Trodrige committed:
cleaned showReport function, and sending data to view

Trodrige committed:
attempt to display rendered data in the view

Trodrige commented:
I have successfully retrieved data based on selected components and send to the view. In the view, when I try to display this data, I get several errors shown below.
1. function that retrieves data
<img width="1280" alt="screen shot 2018-06-23 at 7 22 58 pm" src="https://user-images.githubusercontent.com/16147419/41812452-dec8ae4c-771a-11e8-8563-5d52c3eb8f7e.png">

2. part where I try to display the data. it gives error, "expecting string, object supplied" when there's not print_r() function. I've tried looking into it...
<img width="1278" alt="report_blade_php_ ___sites_librehealth_lh-ehr-laravel-port" src="https://user-images.githubusercontent.com/16147419/41812504-d5b1c2c0-771b-11e8-8a39-d3551f3d5bd3.png">

3. Result from code above
<img width="1280" alt="screen shot 2018-06-23 at 7 32 09 pm" src="https://user-images.githubusercontent.com/16147419/41812516-29b4bb66-771c-11e8-8254-e38184299613.png">

4. Output without print_r() function
<img width="1280" alt="screen shot 2018-06-23 at 7 33 10 pm" src="https://user-images.githubusercontent.com/16147419/41812523-4ff7437a-771c-11e8-9913-80e2b901245a.png">

This error may be due to some very SMALL mistake I've made somewhere. my eyes are open!!!

pri2si17-1997 commented:
Hi @Trodrige ,

From your screenshot, what I guess is you are using object instead of string in blade templating. But this is to be noted that blade escapes string, not object so you are getting this error.

Your error in your foreach statement that you encircled.

What I would suggest is follow this link to see what you actually want to echo : 
https://stackoverflow.com/questions/43217872/laravel-htmlspecialchars-expects-parameter-1-to-be-string-object-given

Other helpful link : 
https://laraveldaily.com/echoing-dd-vs-var_dump-vs-print_r/

Debug the info from above link and see what you are missing. I guess this is sufficient for you to have some more brainstorming and practice some debugging on your code :wink: 

In case you still remain stuck, do let me know. Its just a minor error.

Trodrige committed:
Displaying rendered data

Trodrige commented:
@pri2si17-1997 thanks for the help, but I figured it out. The last commit has a report displayed based on the selected components. @pri2si17-1997 @aethelwulffe @teryhill please do some testing!!! Thanks.

Trodrige commented:
This is the next issue I've been trying to fix. 'occupation' and 'regdate'  have no entry in the database when I try to display '---' for occupation, it fails and outputs the following. I've tested using empty, isset, NULL, "" etc. It still outputs the same thing. @pri2si17-1997 any ideas, Google hasn't been much help. Still searching, though.
<img width="1280" alt="screen shot 2018-06-25 at 6 50 48 pm" src="https://user-images.githubusercontent.com/16147419/41866765-0440ec2c-78a9-11e8-91d4-9e4b9ab43d9e.png"> 

and

<img width="1280" alt="screen shot 2018-06-25 at 6 50 56 pm" src="https://user-images.githubusercontent.com/16147419/41866785-12ac1bd8-78a9-11e8-8b49-9c8453b04de8.png">

aethelwulffe commented:
I don't have the ability/freedom to test at this moment, but is it 
possible that you do not have the optional "show employer data" turned 
on in globals?  That is an optional feature.


On 2018-06-25 13:53, Tigpezeghe Rodrige wrote:
>
> This is the next issue I've been trying to fix. 'occupation' and 
> 'regdate' have no entry in the database when I try to display '---' 
> for occupation, it fails and outputs the following. I've tested using 
> empty, isset, NULL, "" etc. It still outputs the same thing. 
> @pri2si17-1997 <https://github.com/pri2si17-1997> any ideas, Google 
> hasn't been much help. Still searching, though.
> screen shot 2018-06-25 at 6 50 48 pm 
> <https://user-images.githubusercontent.com/16147419/41866765-0440ec2c-78a9-11e8-91d4-9e4b9ab43d9e.png>
>
> and
>
> screen shot 2018-06-25 at 6 50 56 pm 
> <https://user-images.githubusercontent.com/16147419/41866785-12ac1bd8-78a9-11e8-8b49-9c8453b04de8.png>
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub 
> <https://github.com/LibreHealthIO/lh-ehr-laravel-port/pull/1#issuecomment-400039216>, 
> or mute the thread 
> <https://github.com/notifications/unsubscribe-auth/AAhzF8Eb7YC07uMfOoK1Bax2xPAFEnXzks5uASOugaJpZM4UQLrK>.
>

Trodrige commented:
Freedom!!! @aethelwulffe 
I don't get what you mean about the global.

aethelwulffe commented:
In the test data generator, regdate is not a field provided, and occupation is not inserted as a string, which it should be if it is in the test data at all.
UPDATE patient_data SET regdate = '2018-01-01';
UPDATE patient_data SET occupation = 'Test Rat';
That might help.

The global thing was me shooting off the cuff.  The employer data in demographics is a GUI display option, and doubtlessly has nothing to do with your issue.

Trodrige committed:
Add new system feature

Trodrige committed:
Adding ability to save report format. Adding entries to join table

Trodrige commented:
@pri2si17-1997 in order to these recent changes, you have to run 'php artisan module:migrate-refresh ReportGenerator' in the project folder.
Use the 'librereportgenerator.sql' file in the Documentation folder to populate the draggable_components table. BANG!
1. Generate reports
2. Add System Feature
3. Save Report Format
4. Leave your comments!!!

Trodrige committed:
Updating instruction.txt to show Libre EHR Db configurations

pri2si17-1997 commented:
Hi! @Trodrige 

I got your code working, please see attached screenshot : 
![screenshot from 2018-07-02 22-59-53](https://user-images.githubusercontent.com/12952169/42177829-c8103278-7e4b-11e8-95ce-9044fda12f89.png)


General Comments : 
----------------------------
[1] Search button is not visible.
[2] Not able to find any draggable componets, probably because I haven't imported from your sql file. I can see if I import your data. So in this case, if your table is empty, you should have some relevant message there, like 'Oops! No components available.'
[3] Why query part is disabled?

Specific Comments:
---------------------------
[1] You should have detailed instruction, on how to make it up. And add what is relevant to you. Like if you do not need base db, just remove its all contents, it should not throw error in your migrate command.

[2] Do not take dump of your specific table. Take dump of database.

[3] I am not able to search my created system feature. Nothing happens.

[4] Is there any specific order of dragging components or it will handle automatically?

[5] I am getting this error, when trying to generate report.
![screenshot from 2018-07-02 23-03-28](https://user-images.githubusercontent.com/12952169/42178214-1ac507a4-7e4d-11e8-9ac5-5f4969eb1754.png)


These are my findings. I would suggest you to modify instruction, considering a new bie with only laptop in his hand is going to build your code and request from @aethelwulffe @teryhill @tmccormi to review it and give their feedback.

pri2si17-1997 commented on Documentation/librereportgenerator.sql:
```diff
@@ -0,0 +1,48 @@
+-- --------------------------------------------------------
+
+--
+-- Dumping data for table `report_formats`
```
Take dump of whole database. What if user do not run migrations or somehow migration fails? He/she do not know your table structure.

Trodrige replied:
I took these when I had no entries for report_formats, and system features in the database. I'll populate those tables with some entries and dump again!

pri2si17-1997 replied:
I meant that you are taking dump of one table `draggable_components`. Instead take dump of whole database `librereportgenerator`. I think now I'm clear.

Trodrige replied:
I get you!

Trodrige replied:
If I export the entire database, then it'll have CREATE TABLE xyz, and migration command takes care of creating tables.
We need only the INSERT sql statements in this file.

pri2si17-1997 commented on Modules/ReportGenerator/Resources/views/index.blade.php:
```diff
@@ -96,31 +33,14 @@
+                                    @foreach ($draggable_components as $draggable_component)
+                                        <div class="col-sm-2 wordwrap draggable" id="{{ $draggable_component->option_id }}">
+                                            <p data-toggle="tooltip" data-placement="top" title="{{ $draggable_component->title }}">
+                                                {{ $draggable_component->title }}
```
I like templating. :+1:

pri2si17-1997 commented on Modules/ReportGenerator/install.txt:
```diff
@@ -0,0 +1,6 @@
+1. Create your own .env file from the .env.example fill
+2. Edit the database variables as for your report generator database connection.
+3. Create a new database with the database name you specified in DB_REPORT_GENERATOR_DATABASE
+4. Run php artisan module:seed ReportGenerator
```
No need to have a separate file. Merge this and instructions.txt to Readme.

pri2si17-1997 commented on app/Console/Commands/CreateDatabase.php:
```diff
@@ -39,67 +40,75 @@ public function __construct()
+         *  @TODO: Uncomment the line below when EHR is in full Laravel mode.
+         *  That is, when the EHR has been completely ported to Laravel.
+         */
+    	 //$this->createDB(env('DB_DATABASE'), env('DB_HOST'), env('DB_PORT'), env('DB_USERNAME'), env('DB_PASSWORD'));
```
I have uncommented this line to make it working.

Trodrige replied:
Should I uncomment and push?

Trodrige commented:
Hello @pri2si17-1997 

**General answers**
1. Hover over the search button. I'll make it colourful if it's trouble!

2. For now, you have to import the draggable components. Subsequently, I'll populate the tables with default draggable components, system features and report formats automatically. For now, let's roll this way. I know it's painful, but I'll take care of this!

3. The query part is an aside from my project. That is not necessary, but we'll implement it after the main report generator we had in mind is up and running.

**Specific answers**
1. I've taken care of the instructions for setting up the database. 

2. I don't think I understand!!! Anyways, this is what I think. You want the librereportgenerator.sql file to have INSERTs for all tables. The report generator will have the opportunity for the user to create all those things. In addition, we'll have default draggable components, system features, and report formats. I've not yet added enough data for the last 2.

3. The search isn't working yet. The next thing on my TODO list is creating a page to edit/delete system feature and reports. The system features also have to show up as menu items in the navbar. Then the report formats under each system feature will be displayed in the list under each menu (system feature) item.

4. There is no specific order of dragging components. The generated report displays in the order the user dragged the components. I'm not sure if this answers your question. 

5. For the error you're getting, please make sure you enter correct details for the main libreehr database in your .env file. The laravel application doesn't see your database. Maybe you've specified a wrong database name. Check port, username and password.

aethelwulffe commented:
Teacher:  "Little boy, what is 24 divided by 8?"
Little Boy:  "I don't know.  What do you think?"
Teacher:  "I don't THINK, I KNOW!"
Little Boy:  "Well I don't think I know either!"

On 2. above (sort of):  The Layout Engine allows users to alter tables like patient_data.  This includes deletions.  That is not great, and should at least have a "blacklist" of fields that cannot be modified through the GUI at very least.  Many installations have been bricked by someone doing a general cleanup and deleting something useless appearing like "genericval2".  It may be worth looking at having a robust enough "report troubleshooter" so that if, after the report is created, if something gets broken, each query involved can be run sequentially to report to the user in plain language where the problem is.

Trodrige commented:
@aethelwulffe I don't think I'm getting you clearly here. Which '2 above' are you referring to?
Let me just explain this here.

A user of the report generator will have default draggable components, system features, and report formats that they can't edit or delete. The users can only edit and/or delete those that they have created. There is no action done in this report generator that will delete data in the patient_data table. NONE. The users of the report generator can only view these data. The only data they have permission to edit/delete are those for the draggable components, system features and report formats that they created.

I don't know if this is what you were referring to, but I think it's something we all should know.

Trodrige committed:
Merging install and instruction files in README file

aethelwulffe commented:
Right.  You can't use the report gen to delete core things.  Other things can delete things.  Then your report tries to run.
-The thought above was to see if there was something simple we could do to let the user know why the report they built and have been using has suddenly stopped working.  They might get an error message that means something to them, depending on how the server is set up.  

-I am just thinking about how everything can fall to pieces.  I am a systems engineer.  That is my job.  😜

Trodrige committed:
dumping test data for features and report formats

Trodrige committed:
updating and deleting system features

Trodrige committed:
Delete report format

Trodrige committed:
Edit report format title and description

Trodrige committed:
View existing report from selected report format

Trodrige committed:
hiding save report format button when showing existing report

Trodrige committed:
Adding pdf_report.php for PDF generation function

Trodrige committed:
sending for columns for use in pdf generation

aethelwulffe commented on public/pdf_report.php:
```diff
@@ -1 +0,0 @@
-<?php echo 'yes it worked!!!'; ?>
```
error_log("Yes, it still does work")
👍

Trodrige committed:
keeping all draggable components in drop box and draggable pool.

aethelwulffe commented:
Component pool, or component ocean?  :)

Trodrige committed:
--amend

Trodrige commented:
https://tigrodrige.wordpress.com/2018/08/06/gsoc-2018-final-submission/

If @pri2si17-1997 and @aethelwulffe can confirm, then I can submit my final evaluation and continue work here.

pri2si17-1997 commented:
Hi @Trodrige ,

![bug_report_gen](https://user-images.githubusercontent.com/12952169/44110619-d08da75c-a01d-11e8-84ae-d010ab63f7c6.png)

Please see point 7 and 8. System features are of same name. So is it desirable? IMO system features should be unique.

Trodrige commented:
@pri2si17-1997 thanks for observing. I'll fix that ASAP.

Trodrige commented:
@pri2si17-1997 @aethelwulffe hope y'all doing great. I moved the work to the main repo in order to work on installation. This is the pr below. just head to _localhost:'port'/modules/report_generator/public/reportgenerator_ while trying to test the PR in the link below.
https://github.com/LibreHealthIO/lh-ehr/pull/1225

Pri, I fixed the duplicate entries for system features and report formats too.

robbyoconnor commented:
Shall this be closed?
